### PR TITLE
fdiv: adjust for RC typeTag implementation

### PR DESCRIPTION
**Pull request by timsnyder-siv**
chipsalliance/rocket-chip#2723 implemented typeTag in rocket.FPU
control signals.  Update fdiv accordingly

**Related issue**: 

**Type of change**: new feature 

**Impact**: new rtl 

**Development Phase**: implementation

**Release Notes**

### CODE REVIEW

**Comment by timsnyder-siv** [comment]
@jerryz123 and @abejgonzalez I'm not sure if this is the way you would want to handle RC now having `typeTagIn` instead of `singleIn` and `typeTagOut` instead of `singleOut`.  You may want your logic to handle the other type tags that are possible.  I was only trying to get Boom to compile with bumped RC.

File: src/main/scala/exu/execution-units/fpu.scala [hunk 1]
```diff
@@ -191,6 +191,7 @@ class FPU(implicit p: Parameters) extends BoomModule with tile.HasFPUParameters
     req.in3 := unbox(io_req.rs3_data, tag, minT)
     when (fp_ctrl.swap23) { req.in3 := req.in2 }
     req.typ := ImmGenTyp(io_req.uop.imm_packed)
+    req.fmt := DontCare //FIXME: this probably should be a new ImmGenFmt() utility?
```
Comment (timsnyder-siv) [inline src/main/scala/exu/execution-units/fpu.scala]: Without assigning `req.fmt` here, I hit:
```
[error] (run-main-0) firrtl.passes.PassExceptions: 
[error] firrtl.passes.CheckInitialization$RefNotInitializedException:  @[fpu.scala 185:19] : [module FPU]  Reference dfma_io_in_bits_req is not fully initialized.
[error]    : dfma_io_in_bits_req.fmt <= VOID
[error] firrtl.passes.CheckInitialization$RefNotInitializedException:  @[fpu.scala 185:19] : [module FPU]  Reference fpiu_io_in_bits_req is not fully initialized.
[error]    : fpiu_io_in_bits_req.fmt <= VOID
[error] firrtl.passes.CheckInitialization$RefNotInitializedException:  @[fpu.scala 185:19] : [module FPU]  Reference sfma_io_in_bits_req is not fully initialized.
[error]    : sfma_io_in_bits_req.fmt <= VOID
```
in https://app.circleci.com/pipelines/github/ucb-bar/chipyard/3315/workflows/c3d06b70-4b7d-4229-bdba-3e86c79a2e25/jobs/44012

Assigning them to `DontCare` gets past that exception but then test `rv64ud-p-fadd` fails (possibly along with others) in `make -C sims/firesim/sim TARGET_CONFIG=DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig run-asm-tests`

File: src/main/scala/exu/execution-units/functional-unit.scala [hunk 2]
```diff
@@ -605,6 +605,7 @@ class IntToFPUnit(latency: Int)(implicit p: Parameters)
   req.in2 := unbox(io_req.rs2_data, tag, None)
   req.in3 := DontCare
   req.typ := ImmGenTyp(io_req.uop.imm_packed)
+  req.fmt := DontCare // FIXME: this may not be the right thing to do here
```
Comment (timsnyder-siv) [inline src/main/scala/exu/execution-units/functional-unit.scala:608]: Without assigning `req.fmt` here, I hit:
```
[error] (run-main-0) firrtl.passes.PassExceptions: 
[error] firrtl.passes.CheckInitialization$RefNotInitializedException:  @[functional-unit.scala 598:17] : [module IntToFPUnit]  Reference req is not fully initialized.
[error]    : req.fmt <= VOID
```
in https://app.circleci.com/pipelines/github/ucb-bar/chipyard/3315/workflows/c3d06b70-4b7d-4229-bdba-3e86c79a2e25/jobs/44012

Assigning them to `DontCare` gets past that exception but then test `rv64ud-p-fadd` fails (possibly along with others) in `make -C sims/firesim/sim TARGET_CONFIG=DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig run-asm-tests`

**Review by timsnyder-siv — COMMENTED** [review]
I pushed changes that will get past the error CY CI hit in https://app.circleci.com/pipelines/github/ucb-bar/chipyard/3315/workflows/c3d06b70-4b7d-4229-bdba-3e86c79a2e25/jobs/44012 but I am unsure of how to populate FPInput.fmt in Boom.

The corresponding RC assignment happens at https://github.com/chipsalliance/rocket-chip/pull/2723/files#diff-cf4747af8cdedbd7c39b7112ee0ab1431f427e100f3405eb7fdc8e32691a8652R832

**Comment by jerryz123** [comment]
Can you try 
```
req.fmt := Mux(tag === S, 0.U, 1.U)
```
Thanks.

**Comment by timsnyder-siv** [comment]
Trying.

**Comment by timsnyder-siv** [comment]
@jerryz123 , there are still some failing tests when I `run-asm-tests` in MIDAS-level simulation under `sims/firesim/sim`.  I did confirm that running on CY dev@8836f84c79772e2cf9b8fc9118fb862ed2cce9a7 the tests pass.

To get the commit-log printed to stdout, do I need to mix in Chisel-time overrides of the parameters at https://github.com/timsnyder-siv/riscv-boom/blob/fpTypeTag/src/main/scala/common/parameters.scala#L99-L101.  Oh, I guess I can mix-in `WithBoomCommitLogPrintf` eh?  I'm going to generate some waves and commitlogs.

**Comment by jerryz123** [comment]
Hmmm. Odd. I'll take a look at this later today. Can you at least list all failing tests here? Thanks.

**Comment by timsnyder-siv** [comment]
Sure.  With the mux patch you suggested above:
```
diff --git a/src/main/scala/exu/execution-units/fpu.scala b/src/main/scala/exu/execution-units/fpu.scala
index f45896095..2222eff61 100644
--- a/src/main/scala/exu/execution-units/fpu.scala
+++ b/src/main/scala/exu/execution-units/fpu.scala
@@ -191,7 +191,7 @@ class FPU(implicit p: Parameters) extends BoomModule with tile.HasFPUParameters
     req.in3 := unbox(io_req.rs3_data, tag, minT)
     when (fp_ctrl.swap23) { req.in3 := req.in2 }
     req.typ := ImmGenTyp(io_req.uop.imm_packed)
-    req.fmt := DontCare //FIXME: this probably should be a new ImmGenFmt() utility?
+    req.fmt := Mux(tag === S, 0.U, 1.U)
 
     val fma_decoder = Module(new FMADecoder)
     fma_decoder.io.uopc := io_req.uop.uopc
diff --git a/src/main/scala/exu/execution-units/functional-unit.scala b/src/main/scala/exu/execution-units/functional-unit.scala
index 6b08d1e2e..bb2768b6b 100644
--- a/src/main/scala/exu/execution-units/functional-unit.scala
+++ b/src/main/scala/exu/execution-units/functional-unit.scala
@@ -605,7 +605,7 @@ class IntToFPUnit(latency: Int)(implicit p: Parameters)
   req.in2 := unbox(io_req.rs2_data, tag, None)
   req.in3 := DontCare
   req.typ := ImmGenTyp(io_req.uop.imm_packed)
-  req.fmt := DontCare // FIXME: this may not be the right thing to do here
+  req.fmt := Mux(tag === S, 0.U, 1.U)
   req.fmaCmd := DontCare
 
   assert (!(io.req.valid && fp_ctrl.fromint && req.in1(xLen).asBool),

```
The tests I see failing are:
```
bash-4.2$ grep -L PASSED output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/*.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64ud-p-fcvt.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64ud-p-move.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64ud-p-structural.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64ud-v-fcvt.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64ud-v-move.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64ud-v-structural.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-p-fadd.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-p-fcvt.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-p-fdiv.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-p-fmadd.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-p-fmin.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-p-move.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-v-fadd.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-v-fcvt.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-v-fdiv.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-v-fmadd.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-v-fmin.out
output/f1/FireSim-DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig-BaseF1Config/rv64uf-v-move.out
```

Also `rv64ud-{p,v}-structural` are still running after a looong time.  I suspect that they have hung somehow.

**Comment by jerryz123** [comment]
I've pushed a branch `fmv_bump_fix` that should fix the issue. https://github.com/riscv-boom/riscv-boom/commit/e1a70afed7de77f6ba9f6e501de71f7f41afc47c

Not an especially clean solution, but sufficient for now.

**Comment by timsnyder-siv** [comment]
Great. I'll merge it into my pr when I get back to a terminal

File: src/main/scala/exu/execution-units/fpu.scala (hunk 1 above)
Comment (timsnyder-siv) [reply src/main/scala/exu/execution-units/fpu.scala]: This is addressed (at least for now) by the commit @jerryz123 made e1a70af

File: src/main/scala/exu/execution-units/functional-unit.scala (hunk 2 above)
Comment (timsnyder-siv) [reply src/main/scala/exu/execution-units/functional-unit.scala:608]: @jerryz123 the tests (`make -C sims/firesim/sim TARGET_CONFIG=DDR3FRFCFSLLC4MB_FireSimLargeBoomConfig run-asm-tests`) pass with `req.fmt := DontCare` here.  Should I remove the FIXME comment?
```suggestion
  req.fmt := DontCare
```

or should we assign `req.fmt` the same way you're assigning it in fpu.scala in e1a70af ?

**Comment by timsnyder-siv** [comment]
@jerryz123 does it make sense to you why prepare-build-environment is faling in CI for me at https://app.circleci.com/pipelines/github/riscv-boom/riscv-boom/1923/workflows/8e0b065e-9917-48f6-95d8-e5c1bfebccc8/jobs/17424 ?

I'm starting to look into the CI scripts to understand and thought I'd ask before I start pushing potential fixes or extra `set -x` for more debug output.

**Comment by timsnyder-siv** [comment]
Looks like `$SERVER` is unset for my CirclCI build.  I don't see where it is being set like [.circleci/defaults.sh](https://github.com/timsnyder-siv/riscv-boom/blob/e1a70afed7de77f6ba9f6e501de71f7f41afc47c/.circleci/defaults.sh#L5-L6) says it is.  I assume it is in the the CirclCI secrets setup in the job and not kept in revision control.

Is there something in the CircleCI setup that prevents PR's from forks from having SERVER set?  I wouldn't be surprised if that were the case to prevent non-UCB folks from being able to run code on SERVER.  Maybe preventing code injection attacks on SERVER is the reason CY CI doesn't run on forks?

@davidbiancolin, thoughts on how to move forward?

Here's how I would move forward if I had boom write access:

1. `git push git@github.com:riscv-boom/riscv-boom.git  e1a70afed7de77f6ba9f6e501de71f7f41afc47c:refs/heads/fpTypeTag` to make a new local branch at the base of this PR
2. Change the target of this pr to `riscv-boom:fpTypeTag` instead of master
3. Bump the boom chipyard pointer.
3. Merge the PR so that it shows as being merged in GH
4. Open a new PR to merge `riscv-boom:fpTypeTag` to `riscv-boom:mater` so that the CI will run.
6. Merge the new PR and delete `fpTypeTag` and `fmv_bump_fix` topic branches

If someone with write permission in boom does step 1, creating the local branch, I can update this PR so it's ready to merge.

**Comment by abejgonzalez** [comment]
Sorry barely got to this. 

We prevented CI running on forks in the past for security reasons (like you mentioned code injection attacks). And yes, `SERVER` is setup in the CircleCI GUI and is not in revision control just to keep it secret and not available to the public (anything added in the CircleCI GUI gets masked in the log with `****`). 

With that being said, we enabled running on forks recently for a couple PRs (including this one)... however, this is still broken for now because CircleCI doesn't pass environment variables from the GUI to forked builds ("These variables will only get set for pushes to branches on the main repository initiated by someone with commit access, not for CircleCI jobs triggered from a forked pull request." - https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/). 

Future work needs to be done to make the forked build process a bit cleaner instead of manually giving people access to the projects. (the documentation I linked points to a way to automate this that I haven't gotten to yet).

For now, Ill go ahead and do step 1 of what you said. I also re-disabled running builds on forks since its broken anyways and just wastes dev's time trying to figure out why (sorry Tim).

**Comment by timsnyder-siv** [comment]
> For now, Ill go ahead and do step 1 of what you said.

Thanks.  I realized that I was missing the `refs/heads/` part of the destination ref and I think that git complains about that when you push a raw sha.  You know how to create a branch at that commit though.   :)

**Comment by timsnyder-siv** [comment]
The `chisel34` topic branch will work since it still exists.  I changed the base.  I guess you guys probably bump the CHIPYARD.hash  after bumping boom in CY so I'll leave that out for now.

@jerryz123  or @abejgonzalez, do one of you guys want to hit the big green button on this PR and then open a new one to merge `riscv-boom:chisel34` to master again so that the CI will actually run and we can finish out ucb-bar/chipyard#742

**Comment by jerryz123** [comment]
Yeah that sounds good. I'll merge this now.

**Comment by timsnyder-siv** [comment]
#512 is the new PR to run the CI and get this stuff into master.

**Pull request merged**